refactor iscsi network facts module, remove external grep call, add unit test

mator opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Refactor network facts iscsi module for AIX and HP-UX , by removing external grep call and parse with python.  Add unit test.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
facts module

##### ADDITIONAL INFORMATION
No user visible changes.

Removing code when module is undefined , since module is being defined for this code to run. Using single function for a repeated code (string search).

mator committed:
remove external grep call and parse with python

bcoca commented on lib/ansible/module_utils/facts/network/iscsi.py:
```diff
@@ -80,23 +80,32 @@ def collect(self, module=None, collected_facts=None):
+                            iscsi_facts['iscsi_iqn'] = line.split()[1].rstrip()
             else:
-                aixcmd = '/usr/sbin/lsattr -E -l iscsi0 | grep initiator_name'
+                aixcmd = '/usr/sbin/lsattr -E -l iscsi0'
```
you use get_bin_path above, i would reuse it for all lsattr calls (and do same for the other utilities below)

mator replied:
i can't reuse it , since it is different code path where `module.get_bin_path()` is not available  
when module is null/undefined

https://github.com/ansible/ansible/blob/b615561bfa07f93e20c6d4499cb9b9d93c2f0568/lib/ansible/module_utils/facts/network/iscsi.py#L81-L83


i'm not sure `module.get_bin_path()` would find `/opt/iscsi/bin/iscsiutil` on HP-UX PATH environment variable (/opt/iscsi/bin not in the default PATH env)... But i could make a fallback, like this:

```
cmd = module.get_bin_path('iscsiutil')
if not cmd: cmd = '/opt/iscsi/bin/iscsiutil'
```

bcoca commented on lib/ansible/module_utils/facts/network/iscsi.py:
```diff
@@ -80,23 +80,32 @@ def collect(self, module=None, collected_facts=None):
                 aixret = subprocess.check_output(aixcmd, shell=True)
                 if aixret[0].isalpha():
-                    iscsi_facts['iscsi_iqn'] = aixret.split()[1].rstrip()
+                    for line in aixret.splitlines():
```
might want to make common function instead of creating the same for loop several times?

mator replied:
done

bcoca reviewed (changes requested):
In general changes look good, but added a couple of notes

mator committed:
use function for repeated code

mator committed:
use module.get_bin_path() for iscsiutil on HPUX

mator committed:
some code opt for HPUX

mator committed:
clean up non-module code, module being defined is a requirement for this code

mator committed:
import get_bin_path() directly and use without module prefix

mator committed:
Add integration tests for AIX and HP-UX

samdoran commented on test/units/module_utils/facts/network/test_iscsi_get_initiator.py:
```diff
@@ -0,0 +1,52 @@
+    inst = iscsi.IscsiInitiatorNetworkCollector()
+
+    mocker.patch('sys.platform', 'aix6')
+    module.run_command = Mock(return_value=(0, LSATTR_OUTPUT, ''))  # (rc, out, err)
```
Another way to do this using the `mocker` fixture is like this:
```suggestion
    mocker.patch.object(module, 'run_command', return_value=(0, LSATTR_OUTPUT, ''))
```

samdoran commented on test/units/module_utils/facts/network/test_iscsi_get_initiator.py:
```diff
@@ -0,0 +1,52 @@
+    assert aix_iscsi_expected == inst.collect(module=module)
+
+    mocker.patch('sys.platform', 'hp-ux')
+    module.run_command = Mock(return_value=(0, ISCSIUTIL_OUTPUT, ''))  # (rc, out, err)
```
Same here:
```suggestion
    mocker.patch.object(module, 'run_command', return_value=(0, ISCSIUTIL_OUTPUT, ''))
```

samdoran commented on lib/ansible/module_utils/facts/network/iscsi.py:
```diff
@@ -79,24 +80,27 @@ def collect(self, module=None, collected_facts=None):
-                    iscsi_facts['iscsi_iqn'] = hpuxret.split(":", 1)[1].rstrip()
+            hpuxcmd = "/opt/iscsi/bin/iscsiutil"
+            # try to find it in the default PATH
+            cmd = get_bin_path('iscsiutil')
```
Rather than doing the `if not cmd` dance, pass additional search paths to `get_bin_path()`:
```suggestion
            cmd = get_bin_path('iscsiutil', opt_dirs=['/opt/iscsi/bin'], required=True)
```

mator committed:
add changelog fragment

samdoran commented on test/units/module_utils/facts/network/test_iscsi_get_initiator.py:
```diff
@@ -0,0 +1,52 @@
+    module = Mock()
+    inst = iscsi.IscsiInitiatorNetworkCollector()
+
+    mocker.patch('sys.platform', 'aix6')
```
I had to patch the calls to `get_bin_path()` since `lsattar` and `iscsiutil` do not exist on my dev machine:
```suggestion
    mocker.patch('sys.platform', 'aix6')
    mocker.patch('ansible.module_utils.facts.network.iscsi.get_bin_path', return_value='/sbin/lsattr')
```

mator replied:
done

samdoran commented on test/units/module_utils/facts/network/test_iscsi_get_initiator.py:
```diff
@@ -0,0 +1,52 @@
+    aix_iscsi_expected = {"iscsi_iqn": "iqn.localhost.hostid.7f000002"}
+    assert aix_iscsi_expected == inst.collect(module=module)
+
+    mocker.patch('sys.platform', 'hp-ux')
```
```suggestion
    mocker.patch('sys.platform', 'hp-ux')
    mocker.patch('ansible.module_utils.facts.network.iscsi.get_bin_path', return_value='/sbin/iscsiutil')
```

mator replied:
done

samdoran committed:
Apply suggestions from code review

Co-Authored-By: mator <matorola@gmail.com>

mator committed:
Apply suggestions from code review #2

Co-Authored-By: Sam Doran <sdoran@redhat.com>

mator committed:
Remove strict requirement on executable to exist for get_bin_path() as
it will allow facts gathering to continue without an error. Almost all
other files under facts do not have "required=True" (except 2 files,
which should be probably fixed). And check return value for
get_bin_path() , before run attempt.

mator committed:
small fix

mator committed:
add check for AIX lsattr run_command return code
